### PR TITLE
[#243] Implement OverTime Control

### PR DESCRIPTION
**Pull request by tonythomas01**
More work required to actually implement the business logic, and show up in `tasks.php` sidebar

### CODE REVIEW

**Comment by jaragunde** [comment]
In your first patch, watch out for:
* Commented debug code in the web services files.
* Copy-pasted comments in DAOs, Actions.
* Copyright years on new files.

In the second patch, there must be an error regarding how the weekly goals value is printed. My own 37 h figure (manually calculated) appears as 0:38. Also check some inline comments.

File: model/facade/action/GetPersonalSummaryByLoginDateAction.php [hunk 1]
```diff
@@ -76,6 +77,42 @@ public function __construct(UserVO $userVO, DateTime $date) {
...
+     * @return mixed
+     * @throws null
+     */
+    private function getWorkedHoursInThisYear() {
+        $firstDayOfYear = date_create($this->date->format('Y-1-01'));
+
+        // We need to get the number of hours worked till date
+        $extraHoursAction = new ExtraHoursReportAction($firstDayOfYear, $this->date, $this->userVO);
+        $results = $extraHoursAction->execute();
+        return $results[1][$this->userVO->getLogin()]['total_hours'];
```
Comment (jaragunde) [inline model/facade/action/GetPersonalSummaryByLoginDateAction.php]: It's better if you modify ExtraHoursReportAction to take an optional $user parameter. That's a quite heavy report and you will be wasting resources if you calculate the report for every user and then discard them.

Comment (jaragunde) [inline model/facade/action/GetPersonalSummaryByLoginDateAction.php]: You are already doing that! My apologies :)

**Comment by jaragunde** [comment]
Goals seem to work fine when they are specified in the 1st January - 31st Dec time period.

On the other hand, if you define goals in a different time period (I've tried to set 1st Aug - 4th Sep) it still works as if the end date was 31st Dec.

File: web/viewUsers.php [hunk 2]
```diff
@@ -1047,6 +1050,146 @@ function loadHistories(login, userId){
...
+        id: 'goalGrid',
+        iconCls: 'silk-building',
+        width: 300,
+        store: goalStore,
+        frame: false,
+        header: false,
+        title: 'Goals',
+        loadMask: true,
+        colModel: goalColModel,
+        delMsg: 'Are you sure you want to delete the selected City History entries?',
```
Comment (jaragunde) [inline web/viewUsers.php]: Watch out for c&p issues like this :)

**Comment by jaragunde** [comment]
I found a use case that doesn't look right. I've set a goal of -40h at the end of the year (Dec, 31st), and assigned a journey of 8h starting on Dec, 26th until Jan, 1st. That means 40h of work, to fulfil a goal of -40h it means the goal must be zero.

The results of the weekly goal vary depending on the date, like this:
* Monday 26th: 40h
* Tue-Fri: 0h
* Sat 31st: 40h

**Comment by jaragunde** [comment]
Now I'm getting different results, but still strange ones. This is my setup:
* Journey: 8h between 1 Aug 2016 and 9 Aug 2017.
* City: some city with empty calendar between 1 Aug 2016 and 9 Aug 2017.
* Goals: 0h between 1 Aug 2016 and 31 Dec 2016.
* Empty user, no tasks saved.

With the setup above, my weekly goal is 39.42h for every week starting on week 31. This is wrong for several reasons:
* Goal on week 31 should be 40h.
* The weekly goal should grow every week we move into the future if the user hasn't added any tasks.

Additionally, there is a crash when jumping to any date earlier than the 1st of August:
```
[Wed Aug 10 18:57:01.467661 2016] [:error] [pid 20429] [client 127.0.0.1:55244] PHP Fatal error:  Call to a member function getInitDate() on null in /var/www/html/phpreport/model/facade/action/GetPersonalSummaryByLoginDateAction.php on line 95, referer: http://localhost/phpreport/web/tasks.php?date=2016-07-27

File: model/facade/action/GetWeeklyGoalSummaryAction.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+/*//Test code
+
+$user = new UserVO();
+$user->setLogin('jaragunde');
+$action = new GetPersonalSummaryByUserIdDateAction($user, date_create('2009-12-01'));
+//var_dump($action);
+$result = $action->execute();
+var_dump($result);
+*/
```
Comment (jaragunde) [inline model/facade/action/GetWeeklyGoalSummaryAction.php]: Remove this c&p test code, please :)

**Comment by jaragunde** [comment]
One more thing, lately you've been amending your changes to the last commit of this set. That will make your patches more difficult to review. I recommend you keep working on the branch adding more patches to implement new aspects of the feature or correct issues in your implementation, except if they are trivial changes (like correcting a typo, renaming a variable, etc.).

**Comment by tonythomas01** [comment]
The latest patch improves the accuracy a lot, and now in @jaragunde test case, I get 40.13 in the first week. That extra shows up during the `workable hours` calculation, not part of our code, though!

File: model/facade/action/GetPersonalSummaryByLoginDateAction.php [hunk 4]
```diff
@@ -103,6 +103,17 @@ private function getWorkableHoursInThisJourneyPeriod() {
...
 
     /**
+     * @return float
+     */
+    private function getNumberOfWeeksInThisJourney() {
+        $initDate = $this->currentJourney->getInitDate();
+        $endDate = $this->currentJourney->getEndDate();
+
+        $interval = $initDate->diff( $endDate );
+        return round($interval->days/7, 1);
```
Comment (jaragunde) [inline model/facade/action/GetPersonalSummaryByLoginDateAction.php]: You are getting a float with one decimal digit, and I think you should get an integer. Shouldn't this be `ceil()`?

File: model/facade/action/GetPersonalSummaryByLoginDateAction.php [hunk 5]
```diff
@@ -103,6 +103,17 @@ private function getWorkableHoursInThisJourneyPeriod() {
...
+    }
```
Comment (jaragunde) [inline model/facade/action/GetPersonalSummaryByLoginDateAction.php]: It looks like you aren't using this function!

File: model/facade/action/GetPersonalSummaryByLoginDateAction.php [hunk 6]
```diff
@@ -122,13 +133,13 @@ private function getWorkedHoursInThisJourneyPeriod() {
     private function getWeeksTillEndOfJourneyPeriod() {
         $lastDayOfJourney = $this->currentJourney->getEndDate();
 
-        $interval = $this->currentJourney->getInitDate()->diff( $lastDayOfJourney );
-        return floor($interval->days/7);
+        $interval = $this->date->diff( $lastDayOfJourney );
+        return round($interval->days/7, 1);
```
Comment (jaragunde) [inline model/facade/action/GetPersonalSummaryByLoginDateAction.php]: Like I said before, I think the proper thing to do is using `ceil()`.

File: model/facade/action/GetWeeklyGoalSummaryAction.php [hunk 7]
```diff
@@ -106,14 +106,4 @@ protected function doExecute() {
...
-/*//Test code
-
-$user = new UserVO();
-$user->setLogin('jaragunde');
-$action = new GetPersonalSummaryByUserIdDateAction($user, date_create('2009-12-01'));
-//var_dump($action);
-$result = $action->execute();
-var_dump($result);
-*/
+}
```
Comment (jaragunde) [inline model/facade/action/GetWeeklyGoalSummaryAction.php]: This is the kind of trivial change I would have amended into the previous commit O:-)

No need to do it now, just take it into account in the future.

**Comment by jaragunde** [comment]
I've retaken one of the use cases from above: I've assigned a journey of 8h starting on Dec, 26th until Jan, 1st. That means 40h of work on week 52 of year 2016.

Now these are the results of the weekly goal for that week:
* Monday: 44:26
* Tuesday: 57:08
* (it keeps increasing)
* Friday: 133:19
* Saturday: 400!

The problem with different values for different days in the same week is probably related with the way you round week calculations, you should be always using `ceil()`.

File: model/facade/action/GetPersonalSummaryByLoginDateAction.php [hunk 8]
```diff
@@ -69,13 +78,69 @@ class GetPersonalSummaryByLoginDateAction extends Action{
...
+    }
+
+    /**
+     * @return float
+     */
+    private function getWeeksTillEndOfJourneyPeriod() {
+        $lastDayOfJourney = $this->currentJourney->getEndDate();
+
+        $interval = $this->date->diff( $lastDayOfJourney );
+        return floor($interval->days/7);
```
Comment (jaragunde) [inline model/facade/action/GetPersonalSummaryByLoginDateAction.php]: I'm getting a 0 weekly goal in the w52 example. If I change this `floor` with `ceil`, it works properly.

File: model/facade/action/GetPersonalSummaryByLoginDateAction.php [hunk 9]
```diff
@@ -69,13 +78,69 @@ class GetPersonalSummaryByLoginDateAction extends Action{
...
+    /**
+     * @return float
+     */
+    private function getNumberOfWeeksInThisJourney() {
+        $initDate = $this->currentJourney->getInitDate();
+        $endDate = $this->currentJourney->getEndDate();
+
+        $interval = $initDate->diff( $endDate );
+        return ceil($interval->days/7);
+    }
```
Comment (jaragunde) [inline model/facade/action/GetPersonalSummaryByLoginDateAction.php]: I can't see this function used anywhere! If you don't need it, please remove it.

**Comment by jaragunde** [comment]
We are getting closer! But there is still a dependency on the journey end date that should not be there. Take [this case](https://github.com/Igalia/phpreport/pull/274#issuecomment-238931747) again as an example. It's still exposing different results depending on the day of the week:

* Tuesday Aug, 2nd: 39:42 h
* Wednesday Aug, 3rd: 40:27 h

Ideally, the result for any day in that week should be 40h.

**Comment by jaragunde** [comment]
Getting closer and closer, but I just noticed something else O:-)

Using [this case](https://github.com/Igalia/phpreport/pull/274#issuecomment-238931747) again. Now the results on week 31 are 40:27 h for every day in the week. As I said before, the result for any day in that week should be 40h. What's more, the value still changes if I move the end date of the journey. If I set the journey end date to 31st Dec 2016, the weekly goal becomes 41:54 h for week 31. The goal was set for December, 31st, the change of the journey end date beyond 2016 should not affect the result.

**Comment by jaragunde** [comment]
We are almost there, now we have an issue when you start entering your hours :)

Using [this case](https://github.com/Igalia/phpreport/pull/274#issuecomment-238931747) again. Enter Aug, 1 and you will see the weekly goal is 40h, that's correct. Now add a task with 8 work hours; when you save it the weekly goal will go down to 39:38 h.

The tasks I enter on the current week should not affect my goal for it. The goal should still say 40h so I know I need to have 40h at the end of my work week on Friday.

**Comment by jaragunde** [comment]
I still can reproduce the problem I just mentioned. BTW, I don't get how the worked hours affect your calculations, I must be missing something in your code :)

**Comment by jaragunde** [comment]
> I still can reproduce the problem I just mentioned.

It was a problem on my side! Now I'm sure I'm running the latest version, and it works.

> BTW, I don't understand how the worked hours affect your calculations, I must be missing something in your code :)

Now I see worked hours don't affect your calculations, but that's a problem because the goal doesn't get updated as you work.

Imagine a very simple case, with a user who has a contract between Aug 15 and Aug 28, and their journey is 8h. That means they have two weeks of work, 40h each. The goal for the first week is 40h and the goal for the second week is 80h.

Now imagine that user works normally on the first week, and fulfils the goal of 40h. The goal for the second week should be 40h, but the application still says it's 80.

**Comment by tonythomas01** [comment]
The latest patch solves the exact problem!

**Comment by jaragunde** [comment]
I found a new corner case, although I admit it's a strange one. This is the situation:
* Journey is 8h between Aug 15 and 28.
* Assigned to some city between those dates too.
* Goal is set to -40h between Aug 15 and 21 (not 28!).

This is how I think it should work:
* On week 33, there are 40 workable hours and goal is -40 at the end of that week. That means the weekly goal should be 0, but it is 40 instead.
* On week 34 there are 40 remaining workable hours and default goals apply, which means it should be 0 at the end of the contract.  That means the weekly goal should be 40, but it is 80 instead.

Keep up the good work! This is almost done.

**Comment by jaragunde** [comment]
When trying to fix the situation above, take into account this scenario: https://github.com/Igalia/phpreport/pull/274#issuecomment-238931747

If you do changes to how the init date is calculated, take into account that scenario should work exactly the same when you remove the goal.

**Comment by jaragunde** [comment]
Regarding this scenario: https://github.com/Igalia/phpreport/pull/274#issuecomment-239465993

I think we could reconsider this situation:

> On week 34 there are 40 remaining workable hours and default goals apply, which means it should be 0 at the end of the contract. That means the weekly goal should be 40, but it is 80 instead.

We could leave the result as 80, and assume that when a goal is over, it resets to the default periods. Currently those are:
* init date: min(journey start, current week)
* end date: journey end

I don't really like to use journey start and end dates if they span longer than the current year, so I would use Jan 1, Dec 31 as limits.

**Pull request merged**